### PR TITLE
NATIVE-11: Add settings screen

### PR DESCRIPTION
**Pull request by maxammann**

### CODE REVIEW

File: src/routes/language/components/ChangeLanguageModal.js [hunk 1]
```diff
@@ -19,25 +20,35 @@ const Wrapper = styled.View`
...
     InteractionManager.runAfterInteractions(() => {
       changeLanguage(city, model.code)
     })
   }
 
   render () {
-    const { theme, languages, availableLanguages, currentLanguage } = this.props
+    const {theme, languages, availableLanguages, city, currentLanguage} = this.props
+
+    if (!languages || !currentLanguage || !city) {
```
[the remark this answers is not in the document]
Comment (klinzo) [reply src/routes/language/components/ChangeLanguageModal.js]: Or at least handle this stuff in the containers.

File: src/routes/search/components/SearchModal.js [hunk 2]
```diff
@@ -19,20 +20,23 @@ const Wrapper = styled.View`
...
+  closeModal: () => void,
+  navigation: NavigationScreenProp<*>
 |}
 
 type StateType = {|
   query: string
 |}
 
 class SearchModal extends React.Component<PropsType, StateType> {
+  static navigationOptions = {}
```
Comment (klinzo) [inline src/routes/search/components/SearchModal.js]: Why is this required?

File: src/routes/search/components/SearchModal.js [hunk 3]
```diff
@@ -19,20 +20,23 @@ const Wrapper = styled.View`
...
-  theme: ThemeType,
-  language: string,
-  cityCode: string,
-  closeModal: () => void
+  navigateToCategory: (cityCode: string, language: string, path: string) => StoreActionType,
+  theme: any,
+  language: string | null,
+  cityCode: string | null,
+  closeModal: () => void,
+  navigation: NavigationScreenProp<*>
```
Comment (klinzo) [inline src/routes/search/components/SearchModal.js]: Why is `navigation` required?

File: src/routes/search/containers/SearchModalContainer.js [hunk 4]
```diff
@@ -7,20 +7,36 @@ import type { StoreActionType } from '../../../modules/app/StoreActionType'
 import { withTheme } from 'styled-components/native'
 import createNavigateToCategory from '../../../modules/app/createNavigateToCategory'
 import SearchModal from '../components/SearchModal'
+import { CategoriesMapModel } from '@integreat-app/integreat-api-client'
+import type { NavigationScreenProp } from 'react-navigation'
 
-const mapStateToProps = (state: StateType, ownProps) => {
+type OwnPropsType = {| navigation: NavigationScreenProp<*> |}
+
+export type PropsType = {|
```
Comment (klinzo) [inline src/routes/search/containers/SearchModalContainer.js]: Shouldn't this be the PropsTypes from `SearchModal`? At least it's a duplicate.

File: src/routes/settings/components/SwitchSetting.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import * as React from 'react'
+import { Switch } from 'react-native'
+
+type PropType = {
+  onValueChange: boolean => void,
+  getCurrentValue: () => boolean
+}
+
+type StateType = {
+  value: boolean
```
Comment (klinzo) [inline src/routes/settings/components/SwitchSetting.js]: This is a duplication of the state. It should be enough to pass the value as prop.

File: src/routes/search/containers/SearchModalContainer.js (hunk 4 above)
Comment (klinzo) [reply src/routes/search/containers/SearchModalContainer.js]: Alright, I see. Many PropTypes now. Maybe rename to ConnectPropTypes. Not sure though.

File: src/routes/settings/components/Settings.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+      const settingsArray = await AsyncStorage.multiGet(settingsKeys)
+
+      const settings = reduce(settingsArray,
+        (accumulator, [key, value]) => {
+          accumulator[key] = JSON.parse(value)
+          return accumulator
+        },
+        {})
+
+      this.setState(state => ({...state, settingsLoaded: true, settings}))
```
Comment (klinzo) [inline src/routes/settings/components/Settings.js]: You can simplify this to `this.setState({settingsLoaded: true, settings})

File: src/routes/settings/components/Settings.js [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    } catch (e) {
+      console.error('Failed to load settings.')
+    }
+  }
+
+  async setSetting (changeSetting: SettingsType => $Shape<SettingsType>) {
+    this.setState(state => {
+      const newSettings = changeSetting(state.settings)
+      return ({...state, settings: {...state.settings, ...newSettings}})
```
Comment (klinzo) [inline src/routes/settings/components/Settings.js]: Could be simplified to `this.setState({settings: {...changeSetting(state.settings)}})`

File: src/routes/settings/components/Settings.js [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+import styled from 'styled-components/native'
+
+import SettingItem from './SettingItem'
+import type { ThemeType } from '../../../modules/theme/constants/theme'
+import type { NavigationScreenProp } from 'react-navigation'
+import { reduce } from 'lodash/collection'
+import type { TFunction } from 'react-i18next'
+
+type PropsType = {
+  navigation: NavigationScreenProp<*>,
```
Comment (klinzo) [inline src/routes/settings/components/Settings.js]: Why is navigation needed?

File: src/routes/settings/container/SettingsContainer.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import type { NavigationScreenProp } from 'react-navigation'
+import { translate } from 'react-i18next'
+import { connect } from 'react-redux'
+import type { StateType } from '../../../modules/app/StateType'
+
+type OwnPropsType = {| navigation: NavigationScreenProp<*> |}
+
+type PropsType = {
+  navigation: NavigationScreenProp<*>
+}
```
Comment (klinzo) [inline src/routes/settings/container/SettingsContainer.js]: Should be strict, and AFAIK it should only contain a `language: string`

File: src/modules/app/components/Navigator.js [hunk 10]
```diff
@@ -60,10 +61,10 @@ export const ExtrasStack = createStackNavigator(
 export const AppStack = createStackNavigator(
   {
     'Dashboard': createNavigationScreen(LayoutedDashboardContainer, defaultHeader),
-    // $FlowFixMe No idea why this fails
```
Comment (Schedulaar) [inline src/modules/app/components/Navigator.js]: ❤️

File: src/routes/settings/components/Settings.js [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+              }
+            }
+          },
+          {
+            title: t('privacyPolicy'),
+            onPress: () => {
+              if (language === 'de') {
+                Linking.openURL('https://integreat-app.de/datenschutz-webseite/')
+              } else {
+                Linking.openURL('https://integreat-app.de/en/privacy-website/')
```
Comment (Schedulaar) [inline src/routes/settings/components/Settings.js]: We might want to extract all these urls to some constants file

File: src/routes/settings/components/SwitchSetting.js (hunk 5 above)
Comment (Schedulaar) [reply src/routes/settings/components/SwitchSetting.js]: I agree. The only purpose of this component is to "cache" the value. Since this component is not even used anywhere I guess, it can be removed...

File: src/routes/settings/container/SettingsContainer.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+const mapStateToProps = (state: StateType) => {
+  return {
+    language: state.cityContent.language
+  }
+}
+
+const ThemedSettings = withTheme<PropsType, _>(translate('settings')(Settings))
+export default connect<PropsType, OwnPropsType, _, _, _, _>(mapStateToProps)(
```
Comment (Schedulaar) [inline src/routes/settings/container/SettingsContainer.js]: Haha, this is super ugly :D but oh well^^

File: src/routes/settings/components/Settings.js (hunk 7 above)
Comment (Schedulaar) [reply src/routes/settings/components/Settings.js]: Not sure about this one... I think React only shallowly merges the state. So you could simplify it to
```js
this.setState(state => {settings: {...state.settings, ...changeSetting(state.settings)}})
```

File: src/routes/settings/components/Settings.js [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+        extraData={this.state.settings}
+        renderItem={this.renderItem}
+        renderSectionHeader={this.renderSectionHeader}
+        ItemSeparatorComponent={this.ThemedItemSeparator}
+        SectionSeparatorComponent={this.ThemedItemSeparator}
+        stickySectionHeadersEnabled={false}
+      />
+    )
+  }
+}
```
Comment (Schedulaar) [inline src/routes/settings/components/Settings.js]: Can you reduce the size of this file?

File: src/routes/settings/components/Settings.js [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+      const newSettings = changeSetting(state.settings)
+      return ({...state, settings: {...state.settings, ...newSettings}})
+    })
+
+    const settingsArray = reduce(changeSetting(this.state.settings),
+      (accumulator, value, key) => {
+        accumulator.push([key, JSON.stringify(value)])
+        return accumulator
+      },
+      [])
```
Comment (Schedulaar) [inline src/routes/settings/components/Settings.js]: With lodash you can do 
```js
const settingsArray = toPairs(mapValues(changeSetting(this.state.settings), value => JSON.stringify(value)))
```
Also you should probably do the AsyncStorage stuff in the callback of `setState`, right? Or is it not possible with promises?
What do you think of https://medium.com/front-end-weekly/async-await-with-react-lifecycle-methods-802e7760d802

File: src/routes/settings/components/Settings.js [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+
+  constructor (props: PropsType) {
+    super(props)
+
+    this.state = {settingsLoaded: false, settings: {errorTracking: false, test: false}}
+
+    this.loadSettings()
+  }
+
+  async loadSettings () {
```
Comment (Schedulaar) [inline src/routes/settings/components/Settings.js]: Probably this should not be part of the settingsScreen, because we would want to load Settings  from other parts of the app as well?

File: src/routes/settings/components/Settings.js [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+
+type SectionType = { title: ?string }
+
+const ItemSeparator = styled.View`
+    background-color: ${props => props.theme.colors.textDecorationColor};;
+    height: ${StyleSheet.hairlineWidth};
+`
+
+const SectionHeaderContainer = styled.View`
+`
```
Comment (Schedulaar) [inline src/routes/settings/components/Settings.js]: remove if not needed.

File: src/routes/settings/components/Settings.js [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+            onPress: () => {
+              if (language === 'de') {
+                Linking.openURL('https://integreat-app.de/datenschutz-webseite/')
+              } else {
+                Linking.openURL('https://integreat-app.de/en/privacy-website/')
+              }
+            }
+          },
+          {
+            title: t('version', {version: '??'})
```
Comment (Schedulaar) [inline src/routes/settings/components/Settings.js]: Add version or todo comment + task

File: src/modules/layout/components/Header.js [hunk 18]
```diff
@@ -103,7 +107,8 @@ class Header extends React.PureComponent<PropsType> {
   }
 
   render () {
-    const { theme } = this.props
+    const { theme } = this.props // TODO: When mering the share button task make sure to remove the mock t function
+    const t = (key: string) => key
```
Comment (Schedulaar) [inline src/modules/layout/components/Header.js]: *reminder*

File: src/routes/language/components/ChangeLanguageModal.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (maxammann) [reply src/routes/language/components/ChangeLanguageModal.js]: We can create a task for this, as this only came up because of fixing the flow types.

But unfortunately that task will not have priority. Probably a hoc would make sense, but only if and only if the flow types work good.

The benefit of doing it in the render method is that it is very simple.

File: src/routes/search/components/SearchModal.js (hunk 2 above)
Comment (maxammann) [reply src/routes/search/components/SearchModal.js]: Just added it for consistency

File: src/routes/settings/components/SwitchSetting.js (hunk 5 above)
Comment (maxammann) [reply src/routes/settings/components/SwitchSetting.js]: Yes this is a left-over. Thanks for pointing out!

File: src/routes/settings/components/Settings.js (hunk 7 above)
Comment (maxammann) [reply src/routes/settings/components/Settings.js]: Both do not work because they would overwrite `settingsLoaded`

You can only set the whole state with `setState`

File: src/routes/settings/components/Settings.js (hunk 8 above)
Comment (maxammann) [reply src/routes/settings/components/Settings.js]: Settings will get the navigation prop because connect and withTheme passes it through. There are still some problems with the flow types but it works oke for now.

File: src/routes/settings/components/Settings.js (hunk 11 above)
Comment (maxammann) [reply src/routes/settings/components/Settings.js]: We could not as these are not reused I see not really a point in doing it.

File: src/routes/settings/container/SettingsContainer.js (hunk 12 above)
Comment (maxammann) [reply src/routes/settings/container/SettingsContainer.js]: Yes it is.. ... complicated

File: src/modules/layout/components/Header.js (hunk 18 above)
Comment (maxammann) [reply src/modules/layout/components/Header.js]: You were faster! Pfff!

Comment (maxammann) [reply src/modules/layout/components/Header.js]: Didn't even review your changes to the share task! :D

File: src/routes/settings/components/Settings.js (hunk 15 above)
Comment (maxammann) [reply src/routes/settings/components/Settings.js]: Hm not sure yet, but I think I'll do this in the sentry task.

File: src/modules/app/components/Navigator.js (hunk 10 above)
Comment (maxammann) [reply src/modules/app/components/Navigator.js]: Took me hours :D

File: src/routes/search/components/SearchModal.js (hunk 2 above)
Comment (klinzo) [reply src/routes/search/components/SearchModal.js]: Consistency with what? Is this common?

File: src/routes/settings/components/Settings.js (hunk 7 above)
Comment (Schedulaar) [reply src/routes/settings/components/Settings.js]: Have you tried? Usually setState merges shallowly!

File: src/routes/search/components/SearchModal.js [hunk 19]
```diff
@@ -19,20 +20,23 @@ const Wrapper = styled.View`
...
-type PropsType = {|
+export type PropsType = {|
   categories: CategoriesMapModel | null,
-  navigateToCategory: (cityCode: string, language: string, path: string) => void,
-  theme: ThemeType,
-  language: string,
-  cityCode: string,
-  closeModal: () => void
+  navigateToCategory: (cityCode: string, language: string, path: string) => StoreActionType,
+  theme: any,
```
Comment (klinzo) [inline src/routes/search/components/SearchModal.js]: Why did you change this to any? Problems with flow?

File: src/routes/search/components/SearchModal.js [hunk 20]
```diff
@@ -19,20 +20,23 @@ const Wrapper = styled.View`
...
 
-type PropsType = {|
+export type PropsType = {|
   categories: CategoriesMapModel | null,
-  navigateToCategory: (cityCode: string, language: string, path: string) => void,
-  theme: ThemeType,
-  language: string,
-  cityCode: string,
-  closeModal: () => void
+  navigateToCategory: (cityCode: string, language: string, path: string) => StoreActionType,
```
Comment (klinzo) [inline src/routes/search/components/SearchModal.js]: There should be a more specific type for this, shouldn't it?

File: src/routes/search/components/SearchModal.js (hunk 2 above)
Comment (klinzo) [reply src/routes/search/components/SearchModal.js]: I mean e.g. in the `Dashboard` component you removed the `navigationOptions` completely.

File: src/routes/search/components/SearchModal.js [hunk 21]
```diff
@@ -61,8 +65,13 @@ class SearchModal extends React.Component<PropsType, StateType> {
       .map(category => ({model: category, subCategories: []}))
   }
 
-  onItemPress = (category: {path: string}) => {
+  onItemPress = (category: { path: string }) => {
```
Comment (klinzo) [inline src/routes/search/components/SearchModal.js]: Should we try to always use strict props again?

File: src/routes/language/components/ChangeLanguageModal.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (maxammann) [reply src/routes/language/components/ChangeLanguageModal.js]: https://issues.integreat-app.de/browse/NATIVE-53

[the remark this answers is not in the document]
Comment (maxammann) [reply src/routes/language/components/ChangeLanguageModal.js]: Old but gold task :D

File: src/routes/search/components/SearchModal.js (hunk 21 above)
Comment (maxammann) [reply src/routes/search/components/SearchModal.js]: We could force this with a linting rule. But there are a lot of places right now where it could be used. 

I think it makes most sense for PropTypes and the state. If we have them there I'm fine.

File: src/routes/settings/components/Settings.js (hunk 14 above)
Comment (maxammann) [reply src/routes/settings/components/Settings.js]: _Note if you do return a promise from the lifecycle methods it will be not be awaited._

So you can return it but it will not make a difference

File: src/routes/search/components/SearchModal.js (hunk 20 above)
Comment (maxammann) [reply src/routes/search/components/SearchModal.js]: I tried to use the specific type but it does not work with navigateToCategory right now
